unstructured/documents/elements.py

```diff
@@ -772,20 +786,61 @@ def __init__(
         self.embeddings: Optional[list[float]] = embeddings
 
         if isinstance(element_id, NoID):
-            # NOTE(robinson) - Cut the SHA256 hex in half to get the first 128 bits
-            element_id = hashlib.sha256(text.encode()).hexdigest()[:32]
-
+            self._id = self.calculate_hash(metadata)
         elif isinstance(element_id, UUID):
-            element_id = str(uuid.uuid4())
+            self._id = uuid.uuid4()
+        elif isinstance(element_id, str):
+            self._id = element_id
+        else:
+            raise ValueError("ID must be a string, UUID, or NoID")
```

Comment:
This is real-work in the constructor that can be relocated to the `.id` property like so:
```python
def __init__(...):
    self._id = element_id
    super().__init__(element_id=self.id, ...)  # if we still need the base class to keep track of that for us

...

@property
def id(self) -> str:
    # -- id is hash when element_id=NoID() (default) on construction. This is recomputed on each reference
    # -- because metadata is sometimes added after element construction.
    if isinstance(self._id, NoID):
       # -- not a lot of sense in storing the hash if we're just going to recompute it next time we need it --
        return str(self._calculate_hash())
    # -- id is globally-unique when element_id=UUID() on construction. This is never recomputed --
    if isinstance(self._id, UUID):
        self._id = uuid.uuid4()
    if isinstance(self._id, uuid.UUID):
        return str(self._id)
    # -- element_id is a `str` on construction when deserializing an element from JSON --
    if isinstance(self._id, str)
        return self._id
    raise ValueError ...
```

This prevents "real-work-in-the-constructor" (worth Googling if you're not familiar with the term).

Also, if we do actually need an `.id` setter, it prevents that code from being duplicated.